### PR TITLE
[fei5045.2.types] Get WBT types working for both typescript and flow

### PR DESCRIPTION
**Pull request by somewhatabstract**
## Summary:
- Updates `flowgen` to a version that uses `{...}` instead of `{[key: string]: any}`, and supports the `readonly` keyword on indexers
- Adds custom flow types for the `harness/types.ts`, `harness/adapters/router.tsx` and `harness/adapters/adapters.ts` files to cope with the differences in how typescript creates mapped objects vs Flow, and to accommodate that some types come from `history` in TypeScript, but `react-router-dom` in Flow.

Note that these are tailored to how webapp does things.

Issue: FEI-5045

## Test plan:
1. `yarn typecheck`
2. `yarn clean && yarn build && yarn build:types && yarn build:flowtypes`
3. In `webapp/services/static`, after running `yarn`; `rm -r node_modules/@khanacademy/wonder-blocks-testing && cp -r ../../wonder-blocks/packages/wonder-blocks-testing node_modules/@khanacademy/wonder-blocks-testing`
4. `yarn flow:restart && yarn flow`

### CODE REVIEW

File: packages/wonder-blocks-testing/src/fixtures/types.js.flow [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
```
Comment (somewhatabstract) [inline packages/wonder-blocks-testing/src/fixtures/types.js.flow:1]: **nit:** I think I can delete this now that I fixed `flowgen`'s object type. I'll check.

File: packages/wonder-blocks-testing/src/harness/__tests__/types.flowtest.tsx [hunk 2]
```diff
@@ -59,6 +59,7 @@ import type {
 
 //> should error if the config type is wrong
 // 45 is not a string
+// @ts-expect-error: Type '45' is not assignable to type 'string'.
```
Comment (somewhatabstract) [inline packages/wonder-blocks-testing/src/harness/__tests__/types.flowtest.tsx:62]: **note:** Both of these cases should have been erroring - but the error suppressions got removed. This adds them back now that the types are working correctly.

File: packages/wonder-blocks-testing/src/harness/adapters/router.tsx [hunk 3]
```diff
@@ -48,7 +47,7 @@ type Config = // The initial location to use.
                     /**
                      * The location to use.
                      */
-                    location: string | Location;
+                    location: LocationDescriptor;
```
Comment (somewhatabstract) [inline packages/wonder-blocks-testing/src/harness/adapters/router.tsx:50]: **note:** This type encompasses the idea of a string or the shape of a location, which I believe is what we need here.

File: packages/wonder-blocks-testing/src/harness/types.js.flow [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * version of a adapter set. In reality, a more specific type will be used
+ * with the harness functions that use more specific definitions of known
+ * adapters. This is just to support the base reality of not knowing.
+ *
+ * Use this on input positions only. Output positions for adapters
+ * should infer their type in most cases to ensure the strongest typing of
+ * the adapters.
+ */
+export type TestHarnessAdapters = {|
+    +[adapterID: string]: TestHarnessAdapter<any>,
```
Comment (somewhatabstract) [inline packages/wonder-blocks-testing/src/harness/types.js.flow:25]: **note:** This `+` is what is different here, as well as the stuff below for extracting a config type from an adapter, which is copied from the old WB source.

File: packages/wonder-blocks-testing/src/fixtures/types.ts [hunk 5]
```diff
@@ -16,14 +16,14 @@ export type GetPropsOptions = {
     logHandler: (name: string) => (...args: Array<any>) => void;
 };
 
-export type FixtureProps<TProps extends Record<any, any>> =
+export type FixtureProps<TProps extends object> =
```
Comment (somewhatabstract) [inline packages/wonder-blocks-testing/src/fixtures/types.ts:19]: **note:** This change, along with my update to `flowgen` means that we get the expected `TProps: {...}` output, which ensures that things just work.

File: packages/wonder-blocks-testing/src/harness/__tests__/types.typestest.tsx [hunk 6]
```diff
@@ -59,6 +59,7 @@ import type {
 
 //> should error if the config type is wrong
 // 45 is not a string
+// @ts-expect-error: Type '45' is not assignable to type 'string'.
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/__tests__/types.typestest.tsx:62]: Thanks for adding suppressions in for these test cases.

File: packages/wonder-blocks-testing/src/harness/adapters/adapters.ts [hunk 7]
```diff
@@ -29,4 +32,4 @@ export const DefaultConfigs: TestHarnessConfigs<typeof DefaultAdapters> = {
     data: data.defaultConfig,
     portal: portal.defaultConfig,
     router: router.defaultConfig,
-};
+} as const;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/adapters/adapters.ts:35]: Nice!

File: packages/wonder-blocks-testing/src/harness/adapters/router.tsx [hunk 8]
```diff
@@ -1,9 +1,10 @@
+// WARNING: If you modify this file you may need to update router.jsx.flow
+// to ensure that the Flow types are still correct.
 import * as React from "react";
 
 import {StaticRouter, MemoryRouter, Route, Switch} from "react-router-dom";
 
-// @ts-expect-error [FEI-5019] - TS2305 - Module '"react-router-dom"' has no exported member 'LocationShape'. | TS2305 - Module '"react-router-dom"' has no exported member 'Location'.
-import type {LocationShape, Location} from "react-router-dom";
+import type {LocationDescriptor} from "history";
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/adapters/router.tsx:7]: Yay!

File: packages/wonder-blocks-testing/src/harness/types.ts [hunk 9]
```diff
@@ -21,24 +23,17 @@ export type TestHarnessAdapter<TConfig> = (
  * the adapters.
  */
 export type TestHarnessAdapters = {
-    [adapterID: string]: TestHarnessAdapter<any>;
+    readonly [adapterID: string]: TestHarnessAdapter<any>;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/types.ts:26]: Thanks for updating this to match the flow override.

File: packages/wonder-blocks-testing/src/harness/types.ts [hunk 10]
```diff
@@ -21,24 +23,17 @@ export type TestHarnessAdapter<TConfig> = (
...
- *
- * This is the `TestHarnessAdapter` equivalent of `React.ElementConfig`.
  */
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export type TestHarnessConfig<TAdapter> = ReturnType<ExtractConfig>;
+export type TestHarnessConfig<TAdapter> = TAdapter extends TestHarnessAdapter<
+    infer TConfig
+>
+    ? TConfig
+    : never;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/types.ts:36]: Nice!

File: packages/wonder-blocks-testing/src/harness/types.ts [hunk 11]
```diff
@@ -47,13 +42,8 @@ export type TestHarnessConfig<TAdapter> = ReturnType<ExtractConfig>;
...
- * in the `Configs` object, rather than indicating any kind of problem.
+ * marked as `TestHarnessAdapter<TConfig>`
  */
-export type TestHarnessConfigs<TAdapters extends TestHarnessAdapters> = ObjMap<
-    TAdapters,
-    ExtractMaybeConfig
->;
+export type TestHarnessConfigs<TAdapters extends TestHarnessAdapters> = {
+    [K in keyof TAdapters]: TestHarnessConfig<TAdapters[K]> | null | undefined;
+};
```
Comment (kevinbarabash) [inline packages/wonder-blocks-testing/src/harness/types.ts:49]: I love TS' mapped types.

**Review by kevinbarabash — APPROVED** [review]
LGTM. 🎉  Thanks for improving the TS types as part of the work to fix the Flow types.

**Pull request merged**